[SPARK-15186][ML][DOCS] Add user guide for generalized linear regression

sethah opened the pull request:
## What changes were proposed in this pull request?

This patch adds a user guide section for generalized linear regression and includes the examples from [#12754](https://github.com/apache/spark/pull/12754).
## How was this patch tested?

Documentation only, no tests required.
## Approach

In general, it is a bit unclear what level of detail ought to be included in the user guide since there is a lot of variability within the current user guide. I tried to give a fairly brief mathematical introduction to GLMs, and cover what types of problems they could be used for. Additionally, I included a brief blurb on the IRLS solver. The input/output columns are given in a table as is found elsewhere in the docs (though, again, these appear rather intermittently in the current docs), as well as a table providing the supported families and their link functions.

sethah committed:
adding GLR user guide section

sethah committed:
adding columns and families to description

sethah committed:
cleanup

sethah commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,197 @@ regression model and extracting model summary statistics.
+  <thead>
+    <tr>
+      <th></th>
+      <th>PDF</th>
```
I am not sure if it's necessary to include these formulas which are easily found via search. I lean towards not including them, but added them initially since they are easy to take out. Feedback appreciated.

yanboliang replied:
I vote to remove the column ```PDF```. If users want to understand the meaning of it, they should also refer other documents.

sethah replied:
Removed.

sethah commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,197 @@ regression model and extracting model summary statistics.
+
+### Input Columns
+
+<table class="table">
```
These tables are in some sections and not others. Not sure what the decision criteria for including or not including them is.

MLnick replied:
Yeah this is a good point. I think we should come up with a clear standard on what goes into the user guide. Do we include input/output columns? parameters?

The MLlib guide often included more detail on parameters in many cases. You've suggested elsewhere that including these makes it more difficult to maintain to ensure the params (and defaults etc) in the code match the use guide - and I tend to agree. These rather live in the API docs (to which we typically link in the ml guide).

I'd probably say that if we go with the "check the API docs for all param details" then we should not bother with documenting the various input/output columns in the user guide, for the same reasons of maintainability.

cc @jkbradley @mengxr @yanboliang @srowen for comment

yanboliang replied:
I'd prefer to not include params here and provide link to the API docs.

sethah replied:
I removed the sections for now. I like the idea of not duplicating information, but I'm open to others' thoughts.

sethah commented:
cc @yanboliang @mengxr If you get a chance could you review this? Trying to get into Spark 2.0.

SparkQA commented:
**[Test build #58653 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58653/consoleFull)** for PR 13139 at commit [`6de88b8`](https://github.com/apache/spark/commit/6de88b80d43720fe9e66c48205937b8fafded8d3).

SparkQA commented:
**[Test build #58653 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58653/consoleFull)** for PR 13139 at commit [`6de88b8`](https://github.com/apache/spark/commit/6de88b80d43720fe9e66c48205937b8fafded8d3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

MLnick commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,197 @@ regression model and extracting model summary statistics.
+  </tbody>
+</table>
+
+### Optimization
```
I think in this section we should reiterate that it only works for lower feature dimensions (<4096). We've mentioned that above, but I think it should be made clear _where_ the constraint applies (i.e. it is actually in the solver).

yanboliang replied:
Yeah. And I think we should also have more description of IRLS such as each step of IRLS is to solve WLS problems with normal equations method. I think I can contribute a separate PR to document IRLS and WLS solver in Spark. Then feel free to refer that directly.

sethah replied:
So, I went ahead and added some more detail on the optimization routine. I made an effort to stress the limitations on numFeatures and to give some explanation as to why. Could you take a look at it? I didn't generate the docs to make sure it looks alright just yet, but I wanted to get that up so it could be reviewed.

MLnick commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,197 @@ regression model and extracting model summary statistics.
+<div class="codetabs">
+
+<div data-lang="scala" markdown="1">
+{% include_example scala/org/apache/spark/examples/ml/GeneralizedLinearRegressionExample.scala %}
```
Other examples usually include a link to the API docs also in this section

sethah replied:
Done.

yanboliang commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,197 @@ regression model and extracting model summary statistics.
+
+When working with data that has a relatively small number of features (< 4096), Spark's GeneralizedLinearRegression interface
+allows for flexible specification of [generalized linear models](https://en.wikipedia.org/wiki/Generalized_linear_model) (GLMs) which can be used for various types of
+prediction problems including linear regression, Poisson regression, logistic regression, and others.
```
I think it's better to document which are Spark's GeneralizedLinearRegression supported.

thunterdb replied:
See comment below. Also, it would be more clear to say that the current implementation can only deal with up to 4096 features, and that trying to use more features will lead to an error.

yanboliang commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,197 @@ regression model and extracting model summary statistics.
+
+Contrasted with linear regression where the output is assumed to follow a Gaussian
+distribution, GLMs are specifications of linear models where the response variable $Y_i$ may take on _any_
+distribution from the [exponential family of distributions](https://en.wikipedia.org/wiki/Exponential_family). 
```
Or document the supported distribution here.

sethah replied:
The familes and link function are documented in the table below. I could move the table, or are you suggesting something somewhat different?

thunterdb replied:
You should put this paragraph first (it explains the purpose of GRL), and mention that all the supported families are listed below.

sethah replied:
I changed it up a bit. Let me know if that flows better.

thunterdb replied:
Ok great, thanks.

yanboliang commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,197 @@ regression model and extracting model summary statistics.
+$$
+
+Spark's generalized linear regression interface also provides summary statistics for diagnosing the
+fit of GLM models, including residuals, p-values, deviances, the Akaike information criterion, and
```
Should we list all statistic summary here to let users know?

sethah replied:
I would prefer not to. Summary statistics are relatively easy to add and so this could change rather frequently. We should let the examples document them.

yanboliang commented:
@sethah I think we should also document the similarities and differences between Spark GLM and R glm/glmnet.

sethah committed:
more specific IRLS section

sethah commented:
@yanboliang @MLnick Thanks for the feedback. For now, I've just addressed the comment about the optimization section. I'll address the other comments in my next commit (very soon!).

SparkQA commented:
**[Test build #58842 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58842/consoleFull)** for PR 13139 at commit [`ce7c55e`](https://github.com/apache/spark/commit/ce7c55e14a76dc85bca51a2563d770e3eac3a2a2).

SparkQA commented:
**[Test build #58842 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58842/consoleFull)** for PR 13139 at commit [`ce7c55e`](https://github.com/apache/spark/commit/ce7c55e14a76dc85bca51a2563d770e3eac3a2a2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58842/
Test PASSed.

sethah committed:
typos and reword IRLS section

SparkQA commented:
**[Test build #58843 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58843/consoleFull)** for PR 13139 at commit [`e0079d0`](https://github.com/apache/spark/commit/e0079d03f279dc68eb19faed6d5cb6823802051a).

SparkQA commented:
**[Test build #58843 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58843/consoleFull)** for PR 13139 at commit [`e0079d0`](https://github.com/apache/spark/commit/e0079d03f279dc68eb19faed6d5cb6823802051a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58843/
Test PASSed.

sethah committed:
address changes

sethah commented:
I addressed the review comments. Please let me know what else there is. Also, @yanboliang I'd be happy to add specific differences between R and Spark if you have some things in mind? Thanks!

SparkQA commented:
**[Test build #58897 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58897/consoleFull)** for PR 13139 at commit [`d828909`](https://github.com/apache/spark/commit/d828909c850490c1f6c87f545fa40c3cb51649f3).

SparkQA commented:
**[Test build #58897 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58897/consoleFull)** for PR 13139 at commit [`d828909`](https://github.com/apache/spark/commit/d828909c850490c1f6c87f545fa40c3cb51649f3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58897/
Test PASSed.

thunterdb commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,148 @@ regression model and extracting model summary statistics.
+<div data-lang="scala" markdown="1">
+Refer to the [Scala API docs](api/scala/index.html#org.apache.spark.ml.regression.GeneralizedLinearRegression) for more details.
+
+{% include_example scala/org/apache/spark/examples/ml/GeneralizedLinearRegressionExample.scala %}
```
I am having an error with jekyll here, can someone verify?

sethah replied:
Odd. It works for me. What is your error?

thunterdb replied:
Looks like it is working now

sethah committed:
restructure opening section

SparkQA commented:
**[Test build #59016 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59016/consoleFull)** for PR 13139 at commit [`264a490`](https://github.com/apache/spark/commit/264a490c4649e621539af4d88d07b8f3fff1dc2a).

SparkQA commented:
**[Test build #59016 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59016/consoleFull)** for PR 13139 at commit [`264a490`](https://github.com/apache/spark/commit/264a490c4649e621539af4d88d07b8f3fff1dc2a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59016/
Test PASSed.

jkbradley commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,154 @@ regression model and extracting model summary statistics.
+[below](#available-families).
+
+**NOTE**: Spark currently only supports up to 4096 features for GLM models, and will throw an exception if this 
+constraint is exceeded. See the [optimization section](#optimization) for more details.
```
Note that, for certain models, you can call LinearRegression or LogisticRegression to use other solvers which support more features.

sethah replied:
Added.

jkbradley commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,154 @@ regression model and extracting model summary statistics.
+**NOTE**: Spark currently only supports up to 4096 features for GLM models, and will throw an exception if this 
+constraint is exceeded. See the [optimization section](#optimization) for more details.
+
+In a GLM the resonse variable $Y_i$ is assumed to be drawn from an exponential family distribution:
```
typo: "response"

jkbradley commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,154 @@ regression model and extracting model summary statistics.
+In a GLM the resonse variable $Y_i$ is assumed to be drawn from an exponential family distribution:
+
+$$
+Y_i \sim f\left(\cdot|\theta_i, \phi, w_i\right)
```
phi should be defined.

jkbradley replied:
Same for any other notation

sethah replied:
Done.

jkbradley commented:
@yanboliang @sethah Could you please reconcile this PR with [https://github.com/apache/spark/pull/13262]?  Either option is OK with me.  If I had to choose, I'd put the optimization stuff in ml-advanced since most users will not need to know it.

@sethah Where are you drawing your notation from?  If it's a source online, could you link to it?

yanboliang commented:
@jkbradley @sethah I vote to put IRLS session in ml-advance, MLlib IRLS can be used for not only GLM but also other optimization problems such as robust regression in the future. So it's more appropriate to be put in the common session shared by all algorithms rather than in a specific one.

sethah commented:
@yanboliang SGTM. I'll update this one.

sethah committed:
remove optimization section and address review

sethah commented:
@jkbradley Thanks for the review! I wasn't sure if you wanted me to link to the reference I'm using in the actual user guide, or just here for reference. I added it to the user guide for now. Let me know if there is anything else.

SparkQA commented:
**[Test build #59203 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59203/consoleFull)** for PR 13139 at commit [`bd30608`](https://github.com/apache/spark/commit/bd306083c312e3e2e86dd5ad09140cebd152f18c).

SparkQA commented:
**[Test build #59203 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59203/consoleFull)** for PR 13139 at commit [`bd30608`](https://github.com/apache/spark/commit/bd306083c312e3e2e86dd5ad09140cebd152f18c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59203/
Test PASSed.

jkbradley commented:
@sethah Thanks for the updates.  You guessed correctly--I meant to add it to the user guide.  Which section of that guide are you following?  E.g., looking at the GLM theory section, they use slightly different notation.  I wonder if we should match Wikipedia instead since that will probably be the most commonly used reference.  What do you think?

sethah commented:
The only difference here and the reference I linked to (I think) is that I replaced `a_i(\theta)` with its typical form, which they mention, `\phi / w_i`. I think matching Wikipedia is a good idea. I will work on translating the notation.

sethah committed:
switch notation

sethah commented:
@jkbradley I updated the notation to fall in line with Wikipedia.

SparkQA commented:
**[Test build #59227 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59227/consoleFull)** for PR 13139 at commit [`e5eb583`](https://github.com/apache/spark/commit/e5eb583721ea9c560e87186bf9bbd1e13e527c84).

SparkQA commented:
**[Test build #59227 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59227/consoleFull)** for PR 13139 at commit [`e5eb583`](https://github.com/apache/spark/commit/e5eb583721ea9c560e87186bf9bbd1e13e527c84).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59227/
Test PASSed.

jkbradley commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,137 @@ regression model and extracting model summary statistics.
+The canonical form of an exponential family distribution is given as:
+
+$$
+f_Y(y|\theta, \tau) = h(y, \tau)\exp{\left( \frac{\theta \cdot T(y) - A(\theta)}{d(\tau)} \right)}
```
T is not defined (and is discarded below when you mention max likelihood)

sethah replied:
So, it seems as though every source on the internet, academic and otherwise, explains GLMs/exponential families differently with different notation and different terminology. My understanding is that GLMs usually work with an exponential family in its "natural" form, which is a transformed version of an even more generic specification of exponential families. Most _every_ resource I find **besides** wikipedia assumes this "natural" form and does not even mention it. So the `T(y)` appears sometimes, but mostly not. I think the updated explanation is correct, but please let me know if you think it could be clearer.

sethah replied:
Update: I reworded it. GLMs require subsets of the exponential family from the "natural exponential family." See [here](http://www.stats.ox.ac.uk/~steffen/teaching/bs2HT9/glim.pdf) and [here](http://www.biostat.umn.edu/~dipankar/bmtry711.11/lecture_11.pdf).

jkbradley commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,137 @@ regression model and extracting model summary statistics.
+A GLM finds the regression coefficients $\vec{\beta}$ which maximize the likelihood function.
+
+$$
+\min_{\vec{\beta}} \mathcal{L}(\vec{\theta}|\vec{y},X) =
```
max

sethah replied:
Done.

jkbradley commented on docs/ml-classification-regression.md:
```diff
@@ -374,6 +374,137 @@ regression model and extracting model summary statistics.
+by
+
+$$
+\theta_i = A'(g^{-1}(\vec{x_i} \cdot \vec{\beta}))
```
Should A' be inverted?

sethah replied:
Good catch!

jkbradley commented:
Thanks!  Just a few comments

sethah committed:
address review comments

SparkQA commented:
**[Test build #59402 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59402/consoleFull)** for PR 13139 at commit [`a00a470`](https://github.com/apache/spark/commit/a00a47044b109f17cf8a818dd8a2d4c138fc4c5a).

SparkQA commented:
**[Test build #59402 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59402/consoleFull)** for PR 13139 at commit [`a00a470`](https://github.com/apache/spark/commit/a00a47044b109f17cf8a818dd8a2d4c138fc4c5a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59402/
Test PASSed.

sethah committed:
rewording intro

SparkQA commented:
**[Test build #59410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59410/consoleFull)** for PR 13139 at commit [`6e7ddd3`](https://github.com/apache/spark/commit/6e7ddd3f9e91b11979ab704438851ea4d8d99f67).

SparkQA commented:
**[Test build #59410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59410/consoleFull)** for PR 13139 at commit [`6e7ddd3`](https://github.com/apache/spark/commit/6e7ddd3f9e91b11979ab704438851ea4d8d99f67).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59410/
Test FAILed.

jkbradley commented:
@sethah Thanks for the careful checks & writing.  After all this, perhaps it was a pretty tall order to write a brief explanation of GLMs!

LGTM pending tests, which I'll rerun now

SparkQA commented:
**[Test build #3025 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3025/consoleFull)** for PR 13139 at commit [`6e7ddd3`](https://github.com/apache/spark/commit/6e7ddd3f9e91b11979ab704438851ea4d8d99f67).

SparkQA commented:
**[Test build #3025 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3025/consoleFull)** for PR 13139 at commit [`6e7ddd3`](https://github.com/apache/spark/commit/6e7ddd3f9e91b11979ab704438851ea4d8d99f67).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented:
Merging with master and branch-2.0
Thanks!
